File: week2/week2.py
# 第一題
def find_and_print(messages, current_station):
    # Define station names
    station = ['Songshan', 'Nanjing Sanmin', 'Taipei Arena', 'Nanjing Fuxing', 'Songjiang Nanjing', 'Zhongshan', 'Beimen', 'Ximen', 'Xiaonanmen', 'Chiang Kai-Shek Memorial Hall', 'Guting', 'Taipower Building', 'Gongguan', 'Wanlong', 'Jingmei', 'Dapinglin', 'Qizhang', 'Xindian City Hall', 'Xindian']
    # Sanitize station names
    sanitized_station = [station_name.replace(' ', '').replace(',', '').replace('.', '') for station_name in station]

    # Sanitize current station
    sanitized_cur_stat = current_station.replace(' ', '').replace(',', '').replace('.', '')

    # Initialize variables
    min_difference = len(station)
    closest_friend = None

    for friend, message in messages.items():
        # Sanitize message
        sanitized_message = message.replace(' ', '').replace(',', '').replace('.', '')

        found_station = False

        # Check if friend is at Xiaobitan directly
        if 'Xiaobitan' in sanitized_message:
            qizhang_index = sanitized_station.index('Qizhang')
            satation2 = sanitized_station.copy()
            satation2.insert(qizhang_index + 1, 'Xiaobitan')

            difference = abs(satation2.index('Xiaobitan') - satation2.index(sanitized_cur_stat))

            if satation2.index(sanitized_cur_stat) > satation2.index('Xiaobitan'):
                difference += 1

            if difference < min_difference:
                min_difference = difference
                closest_friend = friend

            found_station = True

        else:
            # Iterate through sanitized stations
            for station_name in sanitized_station:
                if station_name in sanitized_message:
                    difference = abs(sanitized_station.index(sanitized_cur_stat) - sanitized_station.index(station_name))

                    if difference < min_difference:
                        min_difference = difference
                        closest_friend = friend

                    found_station = True
                    break

        if not found_station:
            continue

    print(closest_friend or 'No Service')

messages = {
    "Leslie": "I'm at home near Xiaobitan station.",
    "Bob": "I'm at Ximen MRT station.",
    "Mary": "I have a drink near Jingmei MRT station.",
    "Copper": "I just saw a concert at Taipei Arena.",
    "Vivian": "I'm at Xindian station waiting for you."
}

find_and_print(messages, "Wanlong")  # print Mary
find_and_print(messages, "Songshan")  # print Copper
find_and_print(messages, "Qizhang")   # print Leslie
find_and_print(messages, "Ximen")     # print Bob
find_and_print(messages, "Xindian City Hall")  # print Vivian


# 第二題
booked_times = {}

def initialize_booked_times(consultants):
    for consultant in consultants:
        booked_times[consultant["name"]] = [False] * 24

def book(consultants, hour, duration, criteria):
    def is_available(consultant, hour, duration):
        booked_slots = booked_times[consultant["name"]]
        for i in range(hour, hour + duration):
            if booked_slots[i]:
                return False
        return True

    available_consultants = [consultant for consultant in consultants if is_available(consultant, hour, duration)]

    if not available_consultants:
        print("No Service")
        return

    available_consultants.sort(key=lambda x: x[criteria], reverse=criteria == "rate")

    selected_consultant = available_consultants[0]
    print(selected_consultant["name"])

    booked_slots = booked_times[selected_consultant["name"]]
    for i in range(hour, hour + duration):
        booked_slots[i] = True


consultants = [
    {"name": "John", "rate": 4.5, "price": 1000},
    {"name": "Bob", "rate": 3, "price": 1200},
    {"name": "Jenny", "rate": 3.8, "price": 800}
]

initialize_booked_times(consultants)

book(consultants, 15, 1, "price")  # Jenny
book(consultants, 11, 2, "price")  # Jenny
book(consultants, 10, 2, "price")  # John
book(consultants, 20, 2, "rate")   # John
book(consultants, 11, 1, "rate")   # Bob
book(consultants, 11, 2, "rate")   # No Service
book(consultants, 14, 3, "price")  # John

# 第三題
def func(*data):
    def get_middle_name(name):
        parts = list(name)  # Splitting into list of characters
        if len(parts) == 2:
            return parts[1]
        elif len(parts) == 4:
            return parts[2]
        else:
            return parts[1]

    middle_names = list(map(get_middle_name, data))

    unique_middle_name = []
    # unique_indices = []

    for index, name in enumerate(middle_names):
        if name not in unique_middle_name:
            unique_middle_name.append(name)
            # unique_indices.append(index)
        elif name in unique_middle_name:
            unique_middle_name.remove(name)

    if unique_middle_name:
        name_has_unique_mid = next((name for name in data if unique_middle_name[0] in name), None)
        print(name_has_unique_mid)
    else:
        print("沒有")

func("彭大牆", "陳王明雅", "吳明") # print 彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花") # print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花") # print 沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆") # print 夏曼藍波安

# 第四題
def get_number(index):
  quotient = index // 3
  reminder = index % 3
  result = quotient * 7 + (reminder * 4)
  print(result)

get_number(1) # print 4
get_number(5) # print 15
get_number(10) # print 25
get_number(30) # print 70

# 第五題
def find(spaces, stat, n):
    min_spaces_car_index = -1
    min_spaces = float('inf')

    for i, x in enumerate(stat):
        x = spaces[i] if x != 0 else 0
        if x >= n:
            if x < min_spaces:
                min_spaces = x
                min_spaces_car_index = i

    print(min_spaces_car_index)

find([3, 1, 5, 4, 3, 2], [0, 1, 0, 1, 1, 1], 2)  # print 5
find([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4)        # print -1
find([4, 6, 5, 8], [0, 1, 1, 1], 4)               # print 2

File: week2/week2.js
// !第一題
function findAndPrint(messages, currentStation) {
  const station = ['Songshan', 'Nanjing Sanmin', 'Taipei Arena', 'Nanjing Fuxing', 'Songjiang Nanjing', 'Zhongshan', 'Beimen', 'Ximen', 'Xiaonanmen', 'Chiang Kai-Shek Memorial Hall', 'Guting', 'Taipower Building', 'Gongguan', 'Wanlong', 'Jingmei', 'Dapinglin', 'Qizhang', 'Xindian City Hall', 'Xindian']
  const sanitizedStation = station.map(stationName => stationName.replace(/[ ,.]/g, '')) //去除每一站名中的空格和標點符號
  const sanitizedCurStat = currentStation.replace(/[ ,.]/g, '')
  let minDifference = station.length //先設定最小差距為所有站名的數量
  let difference = Infinity
  let closestFriend = null

    for (let friend in messages) { //遍歷每個朋友的訊息
      const sanitizedMessage = messages[friend].replace(/[ ,.]/g, '') //去除訊息中的空格和標點符號
      let foundStation = false //設定一個狀態，若訊息中的站名找到配對則停止遍歷sanitizedStation

      //但是若朋友在xiaoBitan，則直接計算與朋友的差距
      if (sanitizedMessage.includes('Xiaobitan')) {
        const qizhangIndex = sanitizedStation.indexOf('Qizhang')
        const satation2 = sanitizedStation.slice(0); // Make a copy of sanitizedStation
        satation2.splice(qizhangIndex + 1, 0, 'Xiaobitan'); // Array Insert Xiaobitan after Qizhang
        difference = Math.abs(satation2.indexOf('Xiaobitan') - satation2.indexOf(sanitizedCurStat))
        if ((satation2.indexOf(sanitizedCurStat))  > satation2.indexOf('Xiaobitan')) { //如果訊息中包含或者當前站名在Qizhang之後
          difference = difference + 1
        }
          if (difference < minDifference) { //如果遍歷的朋友訊息中的站名的差距小於最小差距，則更新最小差距和最接近的朋友
            minDifference = difference
            closestFriend = friend
          }
          foundStation = true
        } else {

      //接著從sanitizedStaion代入每個站名，看是否比對上朋友所在的站
      for (const stationName of sanitizedStation) { //遍歷sanitizedStation中每個站

      if (sanitizedMessage.includes(stationName)) { //如果訊息中包含站名
        difference = Math.abs(sanitizedStation.indexOf(sanitizedCurStat) - sanitizedStation.indexOf(stationName)); //計算與朋友訊息中的站名序號差距
        if (difference < minDifference) { //如果遍歷的朋友訊息中的站名的差距小於最小差距，則更新最小差距和最接近的朋友
          minDifference = difference;
          closestFriend = friend;
        } 
        foundStation = true
        break; // Exit the loop since we found a matching station 

      }else{continue} //如果訊息中不包含站名，則繼續遍歷下一站點

  }

      if (foundStation) { // Check if station is found in the message
        continue; // Move to the next friend，otherwise will continue to the next station 
      }
    }
  }
  console.log(closestFriend)
}

  const messages = {
  "Bob": "I'm at Ximen MRT station.", 
  "Mary": "I have a drink near Jingmei MRT station.",
  "Copper": "I just saw a concert at Taipei Arena.",
  "Leslie": "I'm at home near Xiaobitan station.",
  "Vivian": "I'm at Xindian station waiting for you."
  }

findAndPrint(messages, "Wanlong"); // print Mary
findAndPrint(messages, "Songshan"); // print Copper
findAndPrint(messages, "Qizhang"); // print Leslie
findAndPrint(messages, "Ximen"); // print Bob
findAndPrint(messages, "Xindian City Hall"); // print Vivian

// !第二題

// Define the createBookingSystem function to encapsulate the booking system logic
function createBookingSystem() {
  const bookedTimes = {};

  // Initialize the booked times array for each consultant
  function initializeBookedTimes(consultants) {
    consultants.forEach(consultant => {
      bookedTimes[consultant.name] = new Array(24).fill(false);
      // {Jonn: Array(24).fill(false), Bob: Array(24).fill(false), Jenny: Array(24).fill(false)}
    });
  }

  // Define the book function
  function book(consultants, hour, duration, criteria) {
    // Check if consultant is available for the given hour and duration
    function isAvailable(consultant, hour, duration) {
      const bookedSlots = bookedTimes[consultant.name];
      for (let i = hour; i < hour + duration; i++) {
        if (bookedSlots[i]) {
          return false; // Overlapping booking found
        }
      }
      return true; // No overlapping booking found
    }

    // Filter available consultants
    let availableConsultants = consultants.filter(consultant => {
      // Check if consultant is available for the given hour and duration
      return isAvailable(consultant, hour, duration);
    });

    if (availableConsultants.length === 0) {
      console.log("No Service");
      return;
    }

    availableConsultants.sort((a, b) => {
      // Sort by criteria
      if (criteria === "price") {
        return a.price - b.price;
      } else if (criteria === "rate") {
        return b.rate - a.rate;
      }
    });

    const selectedConsultant = availableConsultants[0];
    console.log(selectedConsultant.name);

    // Update booked times for the selected consultant
    const bookedSlots = bookedTimes[selectedConsultant.name];
    for (let i = hour; i < hour + duration; i++) {
      bookedSlots[i] = true;
    }
  }

  return { initializeBookedTimes, book };
}

// Create a booking system
const bookingSystem = createBookingSystem();

const consultants = [
  { "name": "John", "rate": 4.5, "price": 1000 },
  { "name": "Bob", "rate": 3, "price": 1200 },
  { "name": "Jenny", "rate": 3.8, "price": 800 }
];

//Using the bookingSystem object, initialize the booked times for each consultant
bookingSystem.initializeBookedTimes(consultants);

// Define the book function as a global function
function book(hour, duration, criteria) {
  bookingSystem.book(consultants, hour, duration, criteria);
}

book(15, 1, "price"); // Jenny
book(11, 2, "price"); // Jenny
book(10, 2, "price"); // John
book(20, 2, "rate"); // John
book(11, 1, "rate"); // Bob
book(11, 2, "rate"); // No Service
book(14, 3, "price"); // John

// !第三題

// collects all the arguments passed to the function into an array named data
//array named data is passed to the map function
function func(...data) {
  //take the middle name of each name and store it in the middleNames array
  let middleNames = data.map(name => {
    let parts = name.split(""); // Split the name into words as an array["彭", "大", "牆]
    if (parts.length === 2) {
      return parts[1];
    } else if (parts.length === 4) {
      return parts[2];
    } else {
      return parts[1];
    }
  });

  let uniqueMiddleName = [];

  middleNames.forEach((name) => { //iterate every middle name of every name
    if (!uniqueMiddleName.includes(name)) {
      uniqueMiddleName.push(name);
    } else if (uniqueMiddleName.includes(name)) {
      uniqueMiddleName.splice(uniqueMiddleName.indexOf(name), 1)
    }
  });

  if (uniqueMiddleName.length !== 0) {
    const nameHasUniqueMid = data.find(name => name.includes(uniqueMiddleName[0]));
    console.log(nameHasUniqueMid);
  } else {
    console.log("沒有");
  }
}
func("彭大牆", "陳王明雅", "吳明"); // print 彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花"); // print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花"); // print 沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆"); // print 夏曼藍波安

// !第四題
function getNumber(index) {
  let quotient = Math.floor(index / 3)
  let reminder = index % 3
  let result = quotient * 7 + (reminder * 4)
  console.log(result)
}
getNumber(1); // print 4
getNumber(5); // print 15
getNumber(10); // print 25
getNumber(30); // print 70

// !第五題

function find(spaces, stat, n) {
  let minSpacesCarIndex = -1
  let minSpaces = Infinity
  stat = stat.map((x, i) => {
    x = x === 0 ? 0 : spaces[i]
    if (x >= n) { //x是空位足夠的車廂空位數，i是車廂序號，n是需要空位
      if (x < minSpaces){
        minSpaces = x
        minSpacesCarIndex = i
      }
    }else{
      return undefined
    }
  })
  console.log(minSpacesCarIndex)
}
find([3, 1, 5, 4, 3, 2], [0, 1, 0, 1, 1, 1], 2); // print 5
find([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4); // print -1
find([4, 6, 5, 8], [0, 1, 1, 1], 4); // print 2